rfcs/0006-transaction.md

```diff
@@ -0,0 +1,301 @@
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+type KeyFingerPrint = u32;
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<KingFingerPrint>,
+    read_keys: HashSet<KeyFingerPrint>,
+    committed_seq: Option<u64>
+}
+```
+
+`Oracle` is considered as a global singleton in the DB, and it tracks the recent committed transactions in the `recent_committed_txns` deque. When a transaction commits, it should push itself into the `recent_committed_txns` deque, and the entries in `recent_committed_txns` can be GCed after they are not needed, we'll cover the details on GC later.
```

Comment:
> When a transaction commits, it should push itself into the `recent_committed_txns` deque

we need to do this for all writes, right? not just transactional writes? For non-transactional writes, what does the `TransactionState` struct look like?

Reply:
yes, non transactional writes is also required. the `TransactionState` for non-transactional writes are exactly same as the transactional ones, having sequence numbers and a set of written keys.

Reply:
I had a duplicate question here: https://github.com/slatedb/slatedb/pull/260#discussion_r1860895052

Would it make sense to expand on how non-transactional updates are handled in the RFC?

Reply:
thank you for the suggestion, updated in c105c22, ptal